src/asmcnc/job/job_data.py

```diff
@@ -389,3 +400,209 @@ def post_job_data_update_post_send(self):
+            print(str(traceback.format_exc()))
+
+    def write_to_recovery_file(self, cancel_line):
+        try:
+            # Account for number of lines added in by the software when running file
+            cancel_line -= self.job_recovery_offset
+            self.job_recovery_offset = 0
+
+            with open(self.job_recovery_info_filepath, 'w+') as job_recovery_info_file:
+                job_recovery_info_file.write(self.filename + "\n" + str(cancel_line))
+
+            # Simultaneously update variables
+            self.job_recovery_filepath = self.filename
+            self.job_recovery_cancel_line = cancel_line
+            self.job_recovery_selected_line = -1
+            self.job_recovery_gcode = []
+            self.job_recovery_offset = 0
+        
+        except:
+            print("Could not write recovery info")
+            print(str(traceback.format_exc()))
+
+    def clear_recovery_file(self):
+        try:
+            open(self.job_recovery_info_filepath, 'w').close()
+
+            # Simultaneously update variables
+            self.job_recovery_filepath = ''
+            self.job_recovery_cancel_line = 0
+            self.job_recovery_selected_line = -1
+            self.job_recovery_gcode = []
+            self.job_recovery_offset = 0
+        
+        except:
+            print("Could not clear recovery info")
+            print(str(traceback.format_exc()))
+
+
+    def reset_recovery(self):
+        self.job_recovery_selected_line = -1
+        self.job_recovery_gcode = []
+        self.job_recovery_offset = 0
+
+    def generate_recovery_gcode(self):
+        try:
+            recovery_gcode = []
+
+            # Recover modal gcodes
+
+            # Coordinate System Select
+            coord_system_line = next((s for s in reversed(self.job_gcode[:self.job_recovery_selected_line]) if re.search("G5[4-9]", s)), None)
+            if coord_system_line:
+                recovery_gcode.append(re.search("G5[4-9](\.[1-3])?", coord_system_line).group(0))
+
+            # Plane selection
+            plane_line = next((s for s in reversed(self.job_gcode[:self.job_recovery_selected_line]) if re.search("G1[7-9]", s)), None)
+            if plane_line:
+                recovery_gcode.append(re.search("G1[7-9]", plane_line).group(0))
+
+            # Distance mode
+            distance_line = next((s for s in reversed(self.job_gcode[:self.job_recovery_selected_line]) if re.search("G9[0,1]([A-Z]|\s|$)", s)), None)
+            if distance_line:
+                # Recovery not allowed if G91 is used
+                if re.search("G91([A-Z]|\s|$)", distance_line):
+                    return False, 'This job contains the G91 g-code, so it cannot be recovered.'
+                recovery_gcode.append("G90")
+
+            # Arc IJK distance mode
+            arc_mode_line = next((s for s in reversed(self.job_gcode[:self.job_recovery_selected_line]) if re.search("G9[0,1]\.1", s)), None)
```

Comment:
@Neex101 - we need to confirm what to do on these. G91.1 probably still needs to be binned (as with G91), but maybe we can handle G91.1 ? That may be conditional.

Reply:
@DennisKiselev I think the correct logic would be to throw an error if G91.1 is the last distance mode used, as it's anotther kind of incremental mode for arcs. 

> G91.1 – Incremental distance mode for I and K offsets. G91.1 returns I and K to their default behavior.

For this merge, it might be best to throw an error if trying to recover a job with G90.1 or G2/G3 as last modes as well, as these are arcs and G2/3 and G90.1 need handling together, but I think we could put in the logic for handling that later on if we wanted.

> G90.1 – Absolute distance mode for I and K offsets. When G90.1 is in effect, I and K both must be specified with G02/G03 for the XZ plane or it is an error.

@Neex101 any thoughts on the above?

Reply:
throw error

Reply:
"Job recovery does not currently support arc distance modes. This job contains [G90.1/G91.1], and therefore cannot be recovered."